# Data & AI Challenge  
## Insurance Claims Fraud Detection

Trabajás con datos reales de siniestros de una aseguradora. El objetivo es mostrar tu capacidad de limpieza de datos, análisis de negocio y uso práctico de IA.

---

### Dataset
**Insurance Claims Fraud Data**

- 3 tablas:
  - `insurance_data` (10k filas)
  - `employee_data`
  - `vendor_data`

---

### Entrega

- **Límite:** Domingo 26/04 a las 18:00 hs  
- **Email:** rrhh@teknedatalabs.com  
- **Asunto:** *Challenge Data & AI - [Tu nombre]*

## Funciones

In [21]:
def analizar_nulos(df, nombre_df="DataFrame"):
    """
    Calcula y muestra los nulos y su porcentaje para un DataFrame dado.
    """
    nulos = df.isnull().sum()
    total = len(df)
    porcentaje = round((nulos / total) * 100, 2)
    
    # Crear reporte en un DataFrame
    df_nulos = pd.DataFrame({
        'Columna': nulos.index,
        'Nulos': nulos.values,
        'Porcentaje (%)': porcentaje.values
    })
    
    # Filtrar solo columnas con nulos y ordenar
    reporte = df_nulos[df_nulos['Nulos'] > 0].sort_values(by='Porcentaje (%)', ascending=False)
    
    print(f"--- Análisis de Nulos: {nombre_df} ---")
    if reporte.empty:
        print("¡Excelente! No hay valores nulos en este DataFrame.")
    else:
        display(reporte)
    print("-" * 35 + "\n")



## 1. Data Engineering & Quality (35%)

- Calculá al menos **dos features derivadas temporales**, por ejemplo:
  - días entre `LOSS_DT` y `REPORT_DT`  
  - antigüedad de la póliza al momento del siniestro  

- Entregá código **Python o SQL reproducible y comentado**

---

**Entregable**

- Notebook o script + breve explicación de cada decisión tomada

### 1.1 - Integrá las 3 tablas y describí brevemente el modelo de datos 

  *(qué representa cada una y cómo se relacionan)*

  - Tablas:
  - `insurance_data`
  - `employee_data`
  - `vendor_data`

**Lectura de los datasets**

In [9]:
import pandas as pd
import os

data_path = "dataset"

dfs = {}

for file in os.listdir(data_path):
    if file.endswith(".csv"):
        file_path = os.path.join(data_path, file)
        
        df_name = file.replace(".csv", "")
        
        df = pd.read_csv(file_path)
        dfs[df_name] = df
        
        print(f"\nDataFrame: {df_name}")
        print(f"Cantidad de filas: {df.shape[0]}")
        print("Primeras 5 filas:")
        display(df.head())
        print("Columnas:")
        print(df.info())
        print("---------------------------------\n")

# Acceso directo a cada DataFrame
insurance_data = dfs['insurance_data']
employee_data = dfs['employee_data']
vendor_data = dfs['vendor_data']



DataFrame: employee_data
Cantidad de filas: 1200
Primeras 5 filas:


,AGENT_ID,AGENT_NAME,DATE_OF_JOINING,ADDRESS_LINE1,ADDRESS_LINE2,CITY,STATE,POSTAL_CODE,EMP_ROUTING_NUMBER,EMP_ACCT_NUMBER
0,AGENT00001,Ray Johns,1993-06-05,1402 Maggies Way,NaN,Waterbury Center,VT,5677,34584958,HKUN51252328472585
1,AGENT00002,Angelo Borjon,2005-12-27,414 Tanya Pass,NaN,Panama City,FL,32404,107363763,OPIS19290040088204
2,AGENT00003,Candy Spellman,2003-09-02,606 National Street,#306,Fayetteville,AR,72701,81744097,YSCJ67489688482590
3,AGENT00004,Mary Smith,2004-09-23,235 Hugh Thomas Drive,NaN,Panama City,FL,32404,67563771,ZANG21285355574581
4,AGENT00005,Mildred Diaz,2011-06-21,3426 Broadview Street,NaN,Montgomery,AL,36110,114951317,DZFS82244494451134


Columnas:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   AGENT_ID            1200 non-null   object
 1   AGENT_NAME          1200 non-null   object
 2   DATE_OF_JOINING     1200 non-null   object
 3   ADDRESS_LINE1       1200 non-null   object
 4   ADDRESS_LINE2       179 non-null    object
 5   CITY                1193 non-null   object
 6   STATE               1200 non-null   object
 7   POSTAL_CODE         1200 non-null   int64 
 8   EMP_ROUTING_NUMBER  1200 non-null   int64 
 9   EMP_ACCT_NUMBER     1200 non-null   object
dtypes: int64(2), object(8)
memory usage: 93.9+ KB
None
---------------------------------


DataFrame: insurance_data
Cantidad de filas: 10000
Primeras 5 filas:


,TXN_DATE_TIME,TRANSACTION_ID,CUSTOMER_ID,POLICY_NUMBER,POLICY_EFF_DT,LOSS_DT,REPORT_DT,INSURANCE_TYPE,PREMIUM_AMOUNT,CLAIM_AMOUNT,...,CLAIM_STATUS,INCIDENT_SEVERITY,AUTHORITY_CONTACTED,ANY_INJURY,POLICE_REPORT_AVAILABLE,INCIDENT_STATE,INCIDENT_CITY,INCIDENT_HOUR_OF_THE_DAY,AGENT_ID,VENDOR_ID
0,2020-06-01 00:00:00,TXN00000001,A00003822,PLC00008468,2015-06-23,2020-05-16,2020-05-21,Health,157.13,9000,...,A,Major Loss,Police,0,1,GA,Savannah,4,AGENT00413,VNDR00556
1,2020-06-01 00:00:00,TXN00000002,A00008149,PLC00009594,2018-04-21,2020-05-13,2020-05-18,Property,141.71,26000,...,A,Total Loss,Ambulance,1,0,AL,Montgomery,0,AGENT00769,VNDR00592
2,2020-06-01 00:00:00,TXN00000003,A00003172,PLC00007969,2019-10-03,2020-05-21,2020-05-26,Property,157.24,13000,...,A,Total Loss,Police,0,1,CO,Grand Junction,19,AGENT00883,VNDR00031
3,2020-06-01 00:00:00,TXN00000004,A00007572,PLC00009292,2016-11-29,2020-05-14,2020-05-19,Health,172.87,16000,...,A,Minor Loss,Ambulance,0,0,GA,Savannah,12,AGENT00278,VNDR00075
4,2020-06-01 00:00:00,TXN00000005,A00008173,PLC00000204,2011-12-26,2020-05-17,2020-05-22,Travel,88.53,3000,...,A,Major Loss,Police,0,1,TN,Nashville,18,AGENT00636,VNDR00472


Columnas:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 38 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   TXN_DATE_TIME             10000 non-null  object 
 1   TRANSACTION_ID            10000 non-null  object 
 2   CUSTOMER_ID               10000 non-null  object 
 3   POLICY_NUMBER             10000 non-null  object 
 4   POLICY_EFF_DT             10000 non-null  object 
 5   LOSS_DT                   10000 non-null  object 
 6   REPORT_DT                 10000 non-null  object 
 7   INSURANCE_TYPE            10000 non-null  object 
 8   PREMIUM_AMOUNT            10000 non-null  float64
 9   CLAIM_AMOUNT              10000 non-null  int64  
 10  CUSTOMER_NAME             10000 non-null  object 
 11  ADDRESS_LINE1             10000 non-null  object 
 12  ADDRESS_LINE2             1495 non-null   object 
 13  CITY                      9946 non-null   object 
 1

,VENDOR_ID,VENDOR_NAME,ADDRESS_LINE1,ADDRESS_LINE2,CITY,STATE,POSTAL_CODE
0,VNDR00001,"King, Proctor and Jones",2027 North Shannon Drive,#5,Fayetteville,AR,72703
1,VNDR00002,Garcia Ltd,5701 East Shirley Lane,NaN,Montgomery,AL,36117
2,VNDR00003,Cherry LLC,1217 Cottondale Road,NaN,Montgomery,AL,36109
3,VNDR00004,Mays-Benson,227 West Montgomery Cross Road,#736,Savannah,GA,31406
4,VNDR00005,Wilson PLC,23 North Hill Street,NaN,Nashville,TN,37210


Columnas:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   VENDOR_ID      600 non-null    object
 1   VENDOR_NAME    600 non-null    object
 2   ADDRESS_LINE1  600 non-null    object
 3   ADDRESS_LINE2  97 non-null     object
 4   CITY           594 non-null    object
 5   STATE          600 non-null    object
 6   POSTAL_CODE    600 non-null    int64 
dtypes: int64(1), object(6)
memory usage: 32.9+ KB
None
---------------------------------



Basado en la estructura de las tablas, este es un **modelo de datos relacional** diseñado para gestionar las operaciones de una compañía de seguros enfocada en "Detección de fraude en reclamos". A continuación, describo las tablas y sus interconexiones:

**employee_data**

Almacena información sobre el personal interno de la compañía que gestiona las pólizas o siniestros. Permite identificar quién es el responsable de cada caso.

**Diccionario de Datos: `employee_data`**

| Campo | Tipo de Dato | Descripción |
| :--- | :--- | :--- |
| `AGENT_ID` | object | Identificador único del agente (Clave primaria). |
| `AGENT_NAME` | object | Nombre completo del agente. |
| `DATE_OF_JOINING` | object | Fecha en la que el agente ingresó a la compañía (formato AAAA-MM-DD). |
| `ADDRESS_LINE1` | object | Línea principal de la dirección física del agente. |
| `ADDRESS_LINE2` | object | Línea secundaria de la dirección (ej. número de oficina o departamento). |
| `CITY` | object | Ciudad de residencia/trabajo del agente. |
| `STATE` | object | Estado o provincia de residencia/trabajo. |
| `POSTAL_CODE` | int64 | Código postal de la ubicación del agente. |
| `EMP_ROUTING_NUMBER` | int64 | Número de ruta bancaria del empleado para pagos/nómina. |
| `EMP_ACCT_NUMBER` | object | Número de cuenta bancaria del empleado. |

**insurance_data**

Es la tabla principal que centraliza toda la actividad operativa. Representa los eventos de transacciones/siniestros. Contiene información detallada sobre el cliente, los detalles específicos de la póliza, la naturaleza del incidente (fecha, severidad, tipo), los montos financieros asociados y la relación con otros actores (agentes y proveedores).

**Diccionario de Datos: `insurance_data`**

| Campo | Tipo | Descripción |
| :--- | :--- | :--- |
| `TXN_DATE_TIME` | object | Fecha y hora en que se registró la transacción. |
| `TRANSACTION_ID` | object | Identificador único de la transacción (Clave primaria). |
| `CUSTOMER_ID` | object | Identificador único del cliente. |
| `POLICY_NUMBER` | object | Número de póliza asociada a la transacción. |
| `POLICY_EFF_DT` | object | Fecha de entrada en vigor de la póliza. |
| `LOSS_DT` | object | Fecha en la que ocurrió la pérdida o siniestro. |
| `REPORT_DT` | object | Fecha en la que se reportó el siniestro. |
| `INSURANCE_TYPE` | object | Tipo de seguro (ej. Salud, Propiedad, Viaje). |
| `PREMIUM_AMOUNT` | float64 | Monto de la prima pagada. |
| `CLAIM_AMOUNT` | int64 | Monto total reclamado por el siniestro. |
| `CUSTOMER_NAME` | object | Nombre del asegurado. |
| `ADDRESS_LINE1` | object | Dirección principal del cliente. |
| `ADDRESS_LINE2` | object | Línea de dirección secundaria (opcional). |
| `CITY` | object | Ciudad de residencia del cliente. |
| `STATE` | object | Estado de residencia del cliente. |
| `POSTAL_CODE` | int64 | Código postal del cliente. |
| `SSN` | object | Número de Seguro Social del cliente. |
| `MARITAL_STATUS` | object | Estado civil del cliente. |
| `AGE` | int64 | Edad del cliente. |
| `TENURE` | int64 | Antigüedad del cliente con la compañía. |
| `EMPLOYMENT_STATUS` | object | Situación laboral actual del cliente. |
| `NO_OF_FAMILY_MEMBERS`| int64 | Número de personas en el grupo familiar del asegurado. |
| `RISK_SEGMENTATION` | object | Categoría de riesgo asignada al cliente. |
| `HOUSE_TYPE` | object | Tipo de vivienda (ej. propia, alquilada). |
| `SOCIAL_CLASS` | object | Clase social o nivel socioeconómico. |
| `ROUTING_NUMBER` | int64 | Número de ruta bancaria del cliente. |
| `ACCT_NUMBER` | object | Número de cuenta bancaria del cliente. |
| `CUSTOMER_EDUCATION_LEVEL` | object | Nivel educativo alcanzado por el cliente. |
| `CLAIM_STATUS` | object | Estado actual de la reclamación (ej. Aprobado, Pendiente). |
| `INCIDENT_SEVERITY` | object | Nivel de gravedad del incidente (ej. Menor, Mayor, Total). |
| `AUTHORITY_CONTACTED` | object | Autoridad notificada sobre el incidente (ej. Policía, Ambulancia). |
| `ANY_INJURY` | int64 | Indicador de lesiones físicas (Binario: 0=No, 1=Sí). |
| `POLICE_REPORT_AVAILABLE` | int64 | Disponibilidad de reporte policial (Binario). |
| `INCIDENT_STATE` | object | Estado donde ocurrió el incidente. |
| `INCIDENT_CITY` | object | Ciudad donde ocurrió el incidente. |
| `INCIDENT_HOUR_OF_THE_DAY`| int64 | Hora del día en que sucedió el incidente. |
| `AGENT_ID` | object | ID del agente encargado (Clave foránea hacia `employee_data`). |
| `VENDOR_ID` | object | ID del proveedor asignado (Clave foránea hacia `vendor_data`). |

**vendor_data**

Contiene los detalles de empresas externas (por ejemplo, talleres de reparación, servicios de salud o servicios de emergencia) que prestan servicios relacionados con la resolución de los siniestros o la atención al cliente.

**Diccionario de Datos: `vendor_data`**


| Campo | Tipo de Dato | Descripción |
| :--- | :--- | :--- |
| `VENDOR_ID` | object | Identificador único del proveedor (Clave primaria). |
| `VENDOR_NAME` | object | Nombre comercial o razón social del proveedor. |
| `ADDRESS_LINE1` | object | Dirección física principal del proveedor. |
| `ADDRESS_LINE2` | object | Dirección secundaria (ej. número de local o suite). |
| `CITY` | object | Ciudad donde opera el proveedor. |
| `STATE` | object | Estado donde se ubica el proveedor. |
| `POSTAL_CODE` | int64 | Código postal de la ubicación del proveedor. |

**Relaciones del Modelo de Datos**

Las tablas están conectadas mediante un esquema relacional donde la tabla de transacciones (`insurance_data`) actúa como el eje central, vinculando a los agentes internos y a los proveedores externos.

**Esquema de Relaciones**

* **`insurance_data` (Tabla central/Hechos):** Es el centro de la actividad. Cada registro representa un siniestro o transacción.
* **`employee_data` (Tabla de Dimensión):** Se conecta a `insurance_data` mediante la columna `AGENT_ID`. Cada registro de un siniestro indica qué agente gestionó dicho caso.
* **`vendor_data` (Tabla de Dimensión):** Se conecta a `insurance_data` mediante la columna `VENDOR_ID`. Cada siniestro puede estar asociado a un proveedor externo (taller, hospital, etc.) que prestó un servicio para resolver el reclamo.

**Detalle de las Conexiones**

| Relación | Tabla Principal | Llave Foránea | Tabla Relacionada | Tipo de Relación |
| :--- | :--- | :--- | :--- | :--- |
| **Gestión** | `insurance_data` | `AGENT_ID` | `employee_data` | Muchos a Uno (Muchos siniestros por un agente) |
| **Servicio** | `insurance_data` | `VENDOR_ID` | `vendor_data` | Muchos a Uno (Muchos siniestros por un proveedor) |

**Beneficios del Modelo**

- **Normalización:** En lugar de repetir los datos completos del empleado (nombre, dirección, datos bancarios) en cada uno de los 10,000 registros de siniestros, el sistema usa solo un `AGENT_ID`. Esto ahorra espacio y evita inconsistencias (si un agente cambia de dirección, solo se actualiza en employee_data).

- **Trazabilidad:** Al realizar un JOIN (o merge en pandas) entre estas tablas, puedes responder preguntas analíticas complejas, tales como:

    ¿Qué agentes tienen el mayor volumen de siniestros con fraude detectado?

    ¿Cuál es el costo promedio de los siniestros atendidos por un proveedor específico en una ciudad determinada?


<img src="dataset/modelado2.png" width="800">

**Integración de las 3 tablas**

Antes de hacer la integración de las tablas iremos al punto 2 para chequear duplicados, nulos, valores atípicos. Ya que al realizar el merge entre  y `insurance_data` y `employee_data` vemos que se incrementa la cantidad de registros que teníamos en los siniestros, ésto no debería suceder, ya que cada transacción está asociada a un único agente. Problablemente la tabla de empleados tenga duplicados para un mismo `AGENT_ID`. Lo chequearemos.

In [11]:
# 1. Realizar el primer merge para traer los datos del agente
# Usamos un 'left' join para mantener todas las transacciones de insurance_data
df_completo = pd.merge(
    insurance_data, 
    employee_data, 
    on='AGENT_ID', 
    how='left'
)

# 2. Realizar el segundo merge para traer los datos del proveedor
#df_final = pd.merge(
#    df_completo, 
#    vendor_data, 
#    on='VENDOR_ID', 
#    how='left'
#)

# Opcional: Verificar el número de columnas antes de los merge 
print(f"Total de columnas de insurance_data antes de la integración: {len(insurance_data.columns)}")


# Opcional: Verificar el número de columnas totales
print(f"Total de columnas después de la integración: {len(df_completo.columns)}")

# 3. Ver los primeros registros del dataframe integrado
df_completo.head()



Total de columnas de insurance_data antes de la integración: 38
Total de columnas después de la integración: 47


,TXN_DATE_TIME,TRANSACTION_ID,CUSTOMER_ID,POLICY_NUMBER,POLICY_EFF_DT,LOSS_DT,REPORT_DT,INSURANCE_TYPE,PREMIUM_AMOUNT,CLAIM_AMOUNT,...,VENDOR_ID,AGENT_NAME,DATE_OF_JOINING,ADDRESS_LINE1_y,ADDRESS_LINE2_y,CITY_y,STATE_y,POSTAL_CODE_y,EMP_ROUTING_NUMBER,EMP_ACCT_NUMBER
0,2020-06-01 00:00:00,TXN00000001,A00003822,PLC00008468,2015-06-23,2020-05-16,2020-05-21,Health,157.13,9000,...,VNDR00556,Amy Wangler,1994-01-06,584 Rural Hill Road,NaN,Nashville,TN,37217,35441155,LRZO53254369804934
1,2020-06-01 00:00:00,TXN00000002,A00008149,PLC00009594,2018-04-21,2020-05-13,2020-05-18,Property,141.71,26000,...,VNDR00592,Jessica Montez,2004-05-28,6114 West Glenn Drive,#1,Glendale,AZ,85301,44055451,XBYG14382904542874
2,2020-06-01 00:00:00,TXN00000003,A00003172,PLC00007969,2019-10-03,2020-05-21,2020-05-26,Property,157.24,13000,...,VNDR00031,Howard Smith,2004-08-01,2500 Snowdoun Chambers Road,NaN,Montgomery,AL,36105,55770246,WRXH14597428306333
3,2020-06-01 00:00:00,TXN00000004,A00007572,PLC00009292,2016-11-29,2020-05-14,2020-05-19,Health,172.87,16000,...,VNDR00075,Patricia Tucker,2007-06-12,313 Lone Oak Drive,NaN,Norman,OK,73071,87652976,YEPN16077902472499
4,2020-06-01 00:00:00,TXN00000005,A00008173,PLC00000204,2011-12-26,2020-05-17,2020-05-22,Travel,88.53,3000,...,VNDR00472,Brenda Bean,2010-01-19,120 Wells Avenue,NaN,Glen Burnie,MD,21061,76119019,HZBK86544578578547


In [12]:
# 1. Agrupar por TRANSACTION_ID y contar cuántos AGENT_ID únicos tiene cada uno
agentes_por_transaccion = insurance_data.groupby('TRANSACTION_ID')['AGENT_ID'].nunique()

# 2. Filtrar transacciones con más de un agente asignado
transacciones_multiples_agentes = agentes_por_transaccion[agentes_por_transaccion > 1]

# 3. Mostrar resultados
if not transacciones_multiples_agentes.empty:
    print(f"Alerta: Se encontraron {len(transacciones_multiples_agentes)} transacciones con más de un agente.")
    # Si quieres ver los IDs afectados:
    print(transacciones_multiples_agentes)
else:
    print("Todo correcto: Cada transacción está asociada a un único agente.")

Todo correcto: Cada transacción está asociada a un único agente.


### 1.2 Documentá **3 problemas de calidad** que encontraste:
  - nulls  
  - tipos incorrectos  
  - outliers  
  - inconsistencias  
  - y cómo los resolviste  



#### Nulos

Chequeamos la existencia de valores nulos en cada tabla.

**`employee_data`**

In [22]:
analizar_nulos(employee_data, "Employee Data")

--- Análisis de Nulos: Employee Data ---


,Columna,Nulos,Porcentaje (%)
4,ADDRESS_LINE2,1021,85.08
5,CITY,7,0.58


-----------------------------------



**Insights**

- `ADDRESS_LINE2`: 85.08% nulos. Al ya contar con valores no nulos en `ADDRESS_LINE1`, tener un alto porcentaje de nulos en esta columna es esperable y no indica problema de calidad, podría ser un campo opcional para el empleado. No indica problema de calidad.

Podemos:
    - dejar como NaN
    - reemplazar por "Sin información"
    - ó eliminar la columna, ya que no aporta valor estadístico. No aporta valor predictivo relevante para fraude ni análisis de negocio. (**Acción elegida**)

- `CITY`: 0.58% nulos

In [ ]:
analizar_nulos(insurance_data, "Insurance Data")
analizar_nulos(vendor_data, "Vendor Data")